gems/pending/Scvmm/miq_hyperv_disk.rb

```diff
@@ -0,0 +1,176 @@
+    @block_size  = 4096
+    @file_size   = 0
+    @block_cache = LruHash.new(DEF_BLOCK_CACHE_SIZE)
+    @cache_hits  = {}
+    @cache_misses = {}
+    @total_read_execution_time = @total_copy_from_remote_time = 0
+  end
+
+  def open(vm_disk)
+    @virtual_disk  = vm_disk
+    @file_offset   = 0
+    stat_script = <<-STAT_EOL
+    (Get-Item "#{@virtual_disk}").length
+STAT_EOL
+    file_size, _stderr   = @parser.parse_single_powershell_value(@winrm.run_powershell_script(stat_script))
+    @file_size           = file_size.to_i
+    @end_byte_addr       = @file_size - 1
+    @size_in_blocks, rem = @file_size.divmod(@block_size)
+    @size_in_blocks += 1 if rem > 0
+    @lba_end = @size_in_blocks - 1
+  end
+
+  def size
+    @file_size
+  end
+
+  def close
+    hit_or_miss if DEBUG_CACHE_STATS
+    @file_offset   = 0
+    @connection    = @winrm = nil
+  end
+
+  def hit_or_miss
+    $log.debug "\nmiq_hyperv_disk cache hits:"
+    @cache_hits.keys.sort.each do |block|
+      $log.debug "block #{block} - #{@cache_hits[block]}"
+    end
+    $log.debug "\nmiq_hyperv_disk cache misses:"
+    @cache_misses.keys.sort.each do |block|
+      $log.debug "block #{block} - #{@cache_misses[block]}"
+    end
+    $log.debug "Total time spent copying reads from remote system is #{@total_copy_from_remote_time}"
+    $log.debug "Total time spent transferring and decoding reads on local system is #{@total_read_execution_time - @total_copy_from_remote_time}"
+    $log.debug "Total time spent processing remote reads is #{@total_read_execution_time}"
+  end
+
+  def seek(offset, whence = IO::SEEK_SET)
+    $log.debug "miq_hyperv_disk.seek(#{offset})"
+    case whence
+    when IO::SEEK_CUR
+      @file_offset += offset
+    when IO::SEEK_END
+      @file_offset = @end_byte_addr + offset
+    when IO::SEEK_SET
+      @file_offset = offset
+    end
+    @file_offset
+  end
+
+  def read(size)
+    $log.debug "miq_hyperv_disk.read(#{size})"
+    return nil if @file_offset >= @file_size
+    size = @file_size - @file_offset if (@file_offset + size) > @file_size
+
+    start_sector, start_offset = @file_offset.divmod(@block_size)
+    end_sector                 = (@file_offset + size - 1) / @block_size
+    number_sectors             = end_sector - start_sector + 1
+
+    @file_offset += size
+    bread_cached(start_sector, number_sectors)[start_offset, size]
+  end
+
+  def bread_cached(start_sector, number_sectors)
+    $log.debug "miq_hyperv_disk.bread_cached(#{start_sector}, #{number_sectors})"
+    @block_cache.keys.each do |block_range|
+      next unless block_range.include?(start_sector) && block_range.include?(start_sector + number_sectors - 1)
+      sector_offset = start_sector - block_range.first
+      buffer_offset = sector_offset * @block_size
+      length        = number_sectors * @block_size
+      if @cache_hits.include?(start_sector)
+        @cache_hits[start_sector] += 1
+      else
+        @cache_hits[start_sector] = 1
+      end
```

Comment:
nitpick: Setting a default value of `0` for the `@cache_hits` hash in `initialize` would eliminate the need for this condition, replacing the entire condition with: `@cache_hits[start_sector] += 1`

No need to change this unless other changes are also needed.